### PR TITLE
Cloners for event menu and layout settings

### PR DESCRIPTION
**Pull request by mic4ael**

### CODE REVIEW

**Comment by ThiefMaster** [comment]
We said we only want one cloner for both. In any case, even if you do want separate checkboxes you do not need two classes.

File: indico/modules/events/layout/clone.py [hunk 1]
```diff
@@ -53,3 +55,44 @@ def clone(self, new_event, options):
...
+            return
+        layout_settings.set(new_event, 'use_custom_menu', layout_settings.get(self.event, 'use_custom_menu'))
+        for menu_entry in menu_entries_for_event(self.event):
+            self._copy_menu_entry(menu_entry, new_event)
+            db.session.add(menu_entry)
+        db.session.flush()
+
+    def _copy_menu_entry(self, menu_entry, event):
+        new_menu_entry = MenuEntry(event_id=event.id)
+        for column in menu_entry.__table__.columns:
```
Comment (ThiefMaster) [inline indico/modules/events/layout/clone.py]: Check what I did in ReminderCloner to iterate over the column attributes properly

File: indico/modules/events/layout/clone.py [hunk 2]
```diff
@@ -53,3 +55,44 @@ def clone(self, new_event, options):
...
+
+class LayoutCloner(EventCloner):
+    def get_options(self):
+        return {'layout': (_('Layout settings'), True, False)}
+
+    def clone(self, new_event, options):
+        if 'layout' not in options:
+            return
+        layout_settings.set_multi(new_event, layout_settings.get_all(self.event))
+        logger.info('Cloned layout settings from event {} to {}'.format(self.event, new_event))
```
Comment (ThiefMaster) [inline indico/modules/events/layout/clone.py]: I think that log message is overkill

File: indico/modules/events/layout/__init__.py [hunk 3]
```diff
@@ -104,3 +104,9 @@ class ImagesFeature(EventFeature):
     @classmethod
     def is_default_for_event(cls, event):
         return event.getType() == 'conference'
+
+
+@signals.event_management.clone.connect
+def _get_layout_and_menu_cloner(event, **kwargs):
+    from indico.modules.events.layout.clone import MenuAndLayoutCloner
+    return MenuAndLayoutCloner(event)
```
Comment (ThiefMaster) [inline indico/modules/events/layout/__init__.py]: Really, MenuAndLayout? Why not just LayoutCloner? Menu is part of the layout module

File: indico/modules/events/layout/clone.py [hunk 4]
```diff
@@ -53,3 +57,31 @@ def clone(self, new_event, options):
             db.session.add(new_image)
             db.session.flush()
             logger.info('Added image during event cloning: {}'.format(new_image))
+
+
+class MenuAndLayoutCloner(EventCloner):
+    def get_options(self):
+        return ({'layout_and_menu': (_('Menu and layout settings'), True, False)}
```
Comment (ThiefMaster) [inline indico/modules/events/layout/clone.py]: layout
"Layout settings and menu customization"?

File: indico/modules/events/layout/clone.py [hunk 5]
```diff
@@ -53,3 +57,31 @@ def clone(self, new_event, options):
...
+                 if self.event.getType() == 'conference' else {})
```
Comment (ThiefMaster) [inline indico/modules/events/layout/clone.py]:
```
if self.event.getType() == 'conference':
    return {}
return {...}
```
Much more redable

File: indico/modules/events/layout/clone.py [hunk 6]
```diff
@@ -53,3 +57,31 @@ def clone(self, new_event, options):
...
+
+    def clone(self, new_event, options):
+        if 'layout_and_menu' not in options or self.event.getType() != 'conference':
```
Comment (ThiefMaster) [inline indico/modules/events/layout/clone.py]: no need to check it again here

File: indico/modules/events/layout/clone.py [hunk 7]
```diff
@@ -53,3 +57,31 @@ def clone(self, new_event, options):
...
+            return
+
+        layout_settings.set_multi(new_event, layout_settings.get_all(self.event))
+        if layout_settings.get(self.event, 'use_custom_menu'):
+            for menu_entry in menu_entries_for_event(self.event):
```
Comment (ThiefMaster) [inline indico/modules/events/layout/clone.py]: Wouldn't it make more sense to query the entries directly?

File: indico/modules/events/layout/clone.py [hunk 8]
```diff
@@ -53,3 +57,31 @@ def clone(self, new_event, options):
...
+
+    def _copy_menu_entry(self, menu_entry, event):
+        base_columns = {column.key for column in inspect(MenuEntry).column_attrs} - {'id', 'event_id'}
+        new_menu_entry = MenuEntry(event_id=event.id, **{col: getattr(menu_entry, col)
+                                                         for col in base_columns})
+        db.session.add(new_menu_entry)
+        db.session.flush()
+        if menu_entry.children:
+            for children in menu_entry.children:
+                child_entry = MenuEntry(event_id=event.id, parent_id=new_menu_entry.id,
```
Comment (ThiefMaster) [inline indico/modules/events/layout/clone.py]: if you append to children you don't need to set the parent id here

File: indico/modules/events/layout/clone.py [hunk 9]
```diff
@@ -53,3 +57,31 @@ def clone(self, new_event, options):
...
+            return
+
+        layout_settings.set_multi(new_event, layout_settings.get_all(self.event))
+        if layout_settings.get(self.event, 'use_custom_menu'):
+            for menu_entry in menu_entries_for_event(self.event):
+                self._copy_menu_entry(menu_entry, new_event)
+            db.session.flush()
+
+    def _copy_menu_entry(self, menu_entry, event):
+        base_columns = {column.key for column in inspect(MenuEntry).column_attrs} - {'id', 'event_id'}
```
Comment (ThiefMaster) [inline indico/modules/events/layout/clone.py]: i think you can safely exclude `parent_id` here

Comment (ThiefMaster) [inline indico/modules/events/layout/clone.py]: also exclude `page_id`

**Comment by ThiefMaster** [comment]
custom pages should be cloned, too

**Comment by ThiefMaster** [comment]
@mic4ael : Can you rebase to master and open a new PR?

File: indico/modules/events/layout/__init__.py [hunk 10]
```diff
@@ -104,3 +104,9 @@ class ImagesFeature(EventFeature):
     @classmethod
     def is_default_for_event(cls, event):
         return event.getType() == 'conference'
+
+
+@signals.event_management.clone.connect
+def _get_layout_and_menu_cloner(event, **kwargs):
```
Comment (ThiefMaster) [inline indico/modules/events/layout/__init__.py]: `_get_layout_cloner`

File: indico/modules/events/layout/clone.py [hunk 11]
```diff
@@ -53,3 +57,36 @@ def clone(self, new_event, options):
...
+                self._copy_menu_entry(menu_entry, new_event)
+            db.session.flush()
+
+    def _copy_menu_entry(self, menu_entry, event):
+        base_columns = {column.key for column in inspect(MenuEntry).column_attrs} - {'id', 'event_id', 'parent_id',
+                                                                                     'page_id'}
+        new_menu_entry = MenuEntry(event_id=event.id, **{col: getattr(menu_entry, col) for col in base_columns})
+        if menu_entry.is_page:
+            new_page = MenuPage(html=menu_entry.page.html)
+            db.session.add(new_page)
```
Comment (ThiefMaster) [inline indico/modules/events/layout/clone.py]: you don't have to add/flush all the time. when you associate the page with the entry, it's added automatically

Comment (ThiefMaster) [inline indico/modules/events/layout/clone.py]: `new_menu_entry.page = MenuPage(html=menu_entry.page.html)` is enough

**Comment by ThiefMaster** [comment]
The default_page_id should be updated, too (it's on `self.event.as_event`)

**Comment by mic4ael** [comment]
Ok, I didn't know about that one.

**Comment by ThiefMaster** [comment]
Superseded by #2036

**Pull request closed without merging**